### PR TITLE
[Feature] Add deduplication and fixed mode for window_funnel

### PR DESCRIPTION
**Pull request by Pslydhh**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Problem Summary(Required) ：

We add two modes for window_funnel
- deduplication: If ordered sequence is A-B-C-B-D, the result only could be A-B-C(3 level).
- fixed: if ordered sequence is A-B-D, the result could be A-B(2 level).

### CODE REVIEW

File: be/src/exprs/agg/window_funnel.h [hunk 1]
```diff
@@ -112,35 +115,231 @@ struct WindowFunnelState {
...
         }
 
-        std::vector<typename TimeType::type> events_timestamp(events_size, -1);
-        for (size_t i = 0; i < events_list.size(); i++) {
-            typename TimeType::type timestamp = (events_list)[i].first;
-            uint8_t event_idx = (events_list)[i].second;
+        if (!mode) {
+            std::vector<typename TimeType::type> events_timestamp(events_size, -1);
+            for (size_t i = 0; i < events_list.size(); i++) {
+                typename TimeType::type timestamp = (events_list)[i].first;
```
Comment (stdpain) [inline be/src/exprs/agg/window_funnel.h:126]: why we use (events_list)

Comment (Pslydhh) [reply be/src/exprs/agg/window_funnel.h:126]: > why we use (events_list)

Your means we should use other data struct(other than std::vector)?

File: be/src/exprs/agg/window_funnel.h [hunk 2]
```diff
@@ -152,6 +351,7 @@ class WindowFunnelAggregateFunction final
 public:
     void update(FunctionContext* ctx, const Column** columns, AggDataPtr __restrict state, size_t row_num) const {
         this->data(state).window_size = ColumnHelper::get_const_value<TYPE_BIGINT>(ctx->get_constant_column(0));
+        this->data(state).mode = ColumnHelper::get_const_value<TYPE_INT>(ctx->get_constant_column(2));
```
[the remark this answers is not in the document]
Comment (Pslydhh) [reply be/src/exprs/agg/window_funnel.h:364]: Now only have 2 modes, we could just ignore others

[the remark this answers is not in the document]
Comment (Pslydhh) [reply be/src/exprs/agg/window_funnel.h:364]: > maybe we should add a limit in FE

Add check for mode in fe

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 1 / 9 (11.11%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|1|9|11.11%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 7 / 9 (77.78%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|7|9|77.78%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 7 / 9 (77.78%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|7|9|77.78%|

**Comment by Pslydhh** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 8 / 11 (72.73%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|8|11|72.73%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 7 / 9 (77.78%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|7|9|77.78%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 7 / 9 (77.78%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|7|9|77.78%|

**Review by stdpain — DISMISSED** [review]

File: be/src/exprs/agg/window_funnel.h [hunk 3]
```diff
@@ -104,9 +107,11 @@ struct WindowFunnelState {
             array.reserve(size * 2 + 2);
             array.emplace_back((int64_t)events_size);
             array.emplace_back((int64_t)sorted);
-            for (int i = 0; i < size; i++) {
-                array.emplace_back((int64_t)events_list[i].first);
-                array.emplace_back((int64_t)events_list[i].second);
+            auto begin = events_list.begin();
```
Comment (satanson) [inline be/src/exprs/agg/window_funnel.h:110]: use it or curr instead of begin

File: be/src/exprs/agg/window_funnel.h [hunk 4]
```diff
@@ -117,30 +122,238 @@ struct WindowFunnelState {
...
+
+        /*
+         * first  of PAIR is the timestamp of event.
+         * second of PAIR is the event_chain_index, It used to eliminate last event chain in dudeplication and order mode.
+         */
+        std::vector<std::pair<typename TimeType::type, int>> events_timestamp(events_size, std::pair(-1, -1));
+        auto begin = ordered_events_list.begin();
+        switch (mode) {
+        // mode: deduplication
+        case 1: {
```
Comment (satanson) [inline be/src/exprs/agg/window_funnel.h:175]: use enum constant instead of literal constant

File: be/src/exprs/agg/window_funnel.h [hunk 5]
```diff
@@ -117,30 +122,238 @@ struct WindowFunnelState {
...
+        // It will update max_level and reduce when encounter condition of deduplication or order mode.
+        int8_t curr_event_level = -1;
+        // It used to index event chain, and monotonic increasing when encounter event_idx = 0.
+        int event_chiain_index = 0;
+
+        /*
+         * first  of PAIR is the timestamp of event.
+         * second of PAIR is the event_chain_index, It used to eliminate last event chain in dudeplication and order mode.
+         */
+        std::vector<std::pair<typename TimeType::type, int>> events_timestamp(events_size, std::pair(-1, -1));
```
Comment (satanson) [inline be/src/exprs/agg/window_funnel.h:171]: so tedious, suggest to use a typedef to define a type alias

File: be/src/exprs/agg/window_funnel.h [hunk 6]
```diff
@@ -41,6 +41,7 @@ struct WindowFunnelState {
     using TimestampEvent = std::pair<typename TimeType::type, uint8_t>;
     mutable std::vector<TimestampEvent> events_list;
```
Comment (satanson) [inline be/src/exprs/agg/window_funnel.h:42]: use typedef instead of typename TimeType::type

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 7 / 9 (77.78%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|7|9|77.78%|

**Comment by kangkaisen** [comment]
```
order: if order sequence is A-B-D, the result could be A-B(2 level).
```
why this is order?  I think this case should be named to continuous.   A-B-D is ordered, but isn't continuous

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 7 / 9 (77.78%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|7|9|77.78%|

**Review by kangkaisen — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Pull request merged**